### PR TITLE
[UI] Collapsible Panels

### PR DESCRIPTION
**Pull request by dawidwoz**
# Problem 

Currently, the view is fixed with the limited capability to focus on some areas of the product, for example, presenting changes to the detail view where we want to make details (right panel) wider than others. 

# Solution
We can tweak CSS to enable collapsible panels. This feature will help with the presentation of new features as well as creates the focus mode for a particular part of the visualisation. It does not change anything for users not interested in changing the current layout. The panels can be hidden by clicking the arrows next to the panels. 
image

The setting is reset to default after you refresh the page below are some examples of the layouts you can create using this CSS tweak. 
image
image
image
image

### CODE REVIEW

**Comment by joglr** [comment]
Hi Dawid.

Thank you for your pull request! 

I don't see the value in resizing the left sidebar - so I suggest we remove that part, unless you have a specific use case for this? If the use case is to have more space available for the rest of the application, I would suggest a simple collapse button instead - something like this: https://dribbble.com/shots/18485896-Collapsible-Sidebar-Interaction
I am thinking something along the existing fullscreen button - just individually controlled for the sidebars instead.

For resizing the visualization, It feels weird that the right sidebar does not move along, when resizing it. Could this be solved with the right sidebar being collapsible as well?

For resizing the right sidebar, it feels really janky that I have to move the mouse in the opposite direction. I understand that this is due to the native resize not being customizable to choose the corner that is resizable. I would like to hear what the resizing, rather than a collapse, brings of value to the user? Is it due to the sidebar being too narrow for the commit view that you have developed? In that case, I think we should investigate alternative solutions to this.

What are your thoughts on this?

**Comment by dawidwoz** [comment]
Hi @joglr,

Thank you for your review. I will try to answer your questions.

`The left sidebar resize`  - exactly, the main scenario is a situation where a user set up everything they wanted and would like to give more space for the application. I agree that the collapse button, as you suggested, would be better. I will be happy to change that. I've been thinking about it I believe the collapsed view should not have the clickable options be just a white strip.  So it would look like this:

image


`It feels weird that the right sidebar does not move along when resizing it. Could this be solved with the right sidebar being collapsible as well?` - True, it is due to the grid layout. It could be fixed by using a `flex` layout instead, however, we use the grid in many places, so I don't want to change it just here.  

`For resizing the right sidebar, it feels really janky that I have to move the mouse in the opposite direction.` - The right panel not moving along and resizing the right panel were accepted by me as the feature was designed to give some more freedom while presenting the tool, so I accepted the current browser behaviour. The alternative would be to use a designed library for it, but it seems a bit overhead for me as we use it in just one place. 


`I would like to hear what the resizing, rather than a collapse, brings of value to the user? Is it due to the sidebar being too narrow for the commit view that you have developed? ` - It was not related to the new commit view at all. It was related to the feedback that 1) the left panel is not that useful when everything was set up but it takes space and 2) In the big screens, the detail panel seem to be too small. 

For example, on the screen in 4K resolution (3840x2160), the view looks likely like this:

image

I agree many users do not use such high resolution but it is used in many conference rooms where this resizing feature cannot be replaced with collapsing panels to adjust the visualisation size. 

It is just an idea so I'm happy to change it to something else or abandon it if you think it's not needed.

**Comment by dawidwoz** [comment]
@joglr I will try to go with the similar solution you have sent here. ([A collapsed view with only icons showing?](https://dribbble.com/shots/19882702-Sidebar-Navigation-1)) 

`The collapsed state was put into the OptionsContext, which is persisted to local storage, so the application remembers the layout chosen.`

No problem!

`A flex layout would be fine for the three major areas` 

I'm trying to experiment with it, but it would be probably not needed for collapsable parts.

`I suggest we start by implementing the collapsible buttons, extending the existing full screen button functionality. Do you agree with this?`

Yes, we can do it now. We can always add resizing when we think it is needed. 

`I would love to collaborate or pair program this if you're up for it, otherwise just try to give it a go and let me know if I can do anything to help.`

Great idea. Now, I'm busy with other projects but we can do some pair programming or GitTruck hackathon in July or August when I am free most of the time. I believe @mircealungu had an idea to organize it in the summer.

**Comment by dawidwoz** [comment]
@joglr I just want to comment to let you know I am still working on this PR and other PRs to backport commit history. The ETA is the beginning of week 27.

**Comment by dawidwoz** [comment]
@joglr I've updated the pull request as we talked. Let me know what you think 🙂

**Review by joglr — APPROVED** [review]
Really nice work! Works really well, I like it. It even works fine along with the fullscreen button.

File: tailwind.config.js [hunk 1]
```diff
@@ -8,6 +8,10 @@ module.exports = {
         blink: "1s linear infinite blink",
         "stroke-pulse": "1s ease-in-out infinite stroke_pulse",
       },
+      inset: {
```
Comment (joglr) [inline tailwind.config.js:11]: I don't quite understand where these insets are used.

**Review by joglr — CHANGES_REQUESTED** [review]
I just did some more testing and I am experiencing some issues with overflow in the right panel.

When commits are expanded, the visualization disappears. It seems like the height of the visualization follows the sidebar for some reason? 

image

**Comment by dawidwoz** [comment]
> I just did some more testing and I am experiencing some issues with overflow in the right panel.
> 
> When commits are expanded, the visualization disappears. It seems like the height of the visualization follows the sidebar for some reason?
> 
> image

@joglr Sorry for the late reply, I was on vacation. Good catch! It should be fixed now.

**Review by joglr — APPROVED** [review]
Hi Dawid, thanks for the update to the PR. I have tested the functionality and it seems to work perfectly. The code is also understandable, so I will approve these changes.

**Comment by mircealungu** [comment]
This is cool stuff @joglr & @dawidwoz :)
